### PR TITLE
cmd/cueckoo: have importpr automatically add "Closes #N"

### PR DESCRIPTION
**Pull request by myitcv**
(see commit message)

### CODE REVIEW

File: cmd/cueckoo/cmd/importpr.go [hunk 1]
```diff
@@ -109,6 +110,13 @@ func importPRDef(c *Command, args []string) error {
 	}
 	log.Printf("fetched PR into branch %q", branchName)
 
+	// Set the branch upstream as the first step. If subsequent commands
+	// (they shouldn't but it can happen) we still need the upstream to have
+	// been set.
```
Comment (mvdan) [inline cmd/cueckoo/cmd/importpr.go:115]: failing to parse the second sentence :)

File: cmd/cueckoo/cmd/importpr.go [hunk 2]
```diff
@@ -140,13 +148,25 @@ func importPRDef(c *Command, args []string) error {
...
-		return err
-	}
 	log.Printf("rebased and squashed on %s", rebaseMsg)
 
 	// TODO: fix up common commit message issues, especially when squashing, in Go code.
-	// TODO: automate adding "Closes #PR as merged.".
+
+	// Add "Closes #PR as merged." Not that running this command will also end
+	// up adding a Change-Id trailer if the user has git commit hooks set for
+	// pre-commit. This is a good thing.
```
Comment (mvdan) [inline cmd/cueckoo/cmd/importpr.go:157]: note that adding the change-id was already happening via the "git commit --amend" below - I imagine you're aware, and that this is for the sake of better docs. just want to flag that this particular behavior should remain unchanged.

File: cmd/cueckoo/cmd/importpr.go [hunk 3]
```diff
@@ -183,3 +203,72 @@ func run(ctx context.Context, name string, args ...string) (string, error) {
...
+	// regexpTrailer represents a token: value pair. Uses a rough translation of
+	// the description about what a token: value pair can be from "man
+	// git-interpret-trailers". This regexp explicitly does not try to
+	//
+	// TODO: refine to be complete. Use "man git-interpret-trailers"?
+	regexpTrailer          = regexp.MustCompile(`^[-a-zA-Z]+\s*:`)
+	regexpContinuationLine = regexp.MustCompile(`^\s+`)
+)
+
+// addCloses adds the message to "Closes #pr as merged." to commit msg. It
```
Comment (mvdan) [inline cmd/cueckoo/cmd/importpr.go:217]:
```suggestion
// addCloses adds the message "Closes #pr as merged." to the commit message msg. It
```

File: cmd/cueckoo/cmd/importpr.go [hunk 4]
```diff
@@ -183,3 +203,72 @@ func run(ctx context.Context, name string, args ...string) (string, error) {
...
+// respects trailers and leaves a newline at the end of the message. Like
+// git it respects the last block of trailers.
+//
+// There is probably a nice package somewhere for parsing the git commit
+// message into the constituent parts: pre-trailers, and trailers. For now
+// we brute force it based on the logic described in "man
+// git-interpret-trailers". If there are trailers we want to insert "Closes
+// #PR as merged." as the last clear line before the trailers. If there are
+// no trailers, it should be the final line in the commit message.
```
Comment (mvdan) [inline cmd/cueckoo/cmd/importpr.go:226]: How about using `git interpret-trailers`? It knows how to separate and parse the trailers, and it also unfolds them, so there's no need to handle continuation lines ourselves.

```
$ git log -1 --format=%B
cue/context: support internal interpreters

This introduces the hooks needed for implementing a WASM
interpreter to be used in CUE. This API, for now, only supports
internal types.

Signed-off-by: Marcel van Lohuizen <mpvl@gmail.com>
Change-Id: Id201419028ca8cbac5883809cdbd53ab7aa440cf
Reviewed-on: https://review.gerrithub.io/c/cue-lang/cue/+/550265
Reviewed-by: Roger Peppe <rogpeppe@gmail.com>
Unity-Result: CUEcueckoo <cueckoo@cuelang.org>
TryBot-Result: CUEcueckoo <cueckoo@cuelang.org>

$ git log -1 --format=%B | git interpret-trailers --only-trailers --only-input --unfold
Signed-off-by: Marcel van Lohuizen <mpvl@gmail.com>
Change-Id: Id201419028ca8cbac5883809cdbd53ab7aa440cf
Reviewed-on: https://review.gerrithub.io/c/cue-lang/cue/+/550265
Reviewed-by: Roger Peppe <rogpeppe@gmail.com>
Unity-Result: CUEcueckoo <cueckoo@cuelang.org>
TryBot-Result: CUEcueckoo <cueckoo@cuelang.org>
```

There's apparently no way to ask it to give us the "body" of the commit message, but we can hack that together by running the helper twice. With `--only-trailers --only-input --unfold` it only gives us the trailers, and with `--only-input --unfold` it gives us everything. If you trim whitespace from both, you can count the number of trailer lines and strip that number of trailing lines from the "all" output to get the body.

Then add the body, the "closes" paragraph, and the unfolded trailers. I realise this is a bit of a refactor, so I'm happy to volunteer to do it if it sounds like a good idea. I'd just feel better about maintaining that code, as we bake in fewer assumptions and there's less chance that we messed up the string handling or logic :)

File: cmd/cueckoo/cmd/importpr.go (hunk 3 above)
Comment (myitcv) [reply cmd/cueckoo/cmd/importpr.go:217]: Done

File: cmd/cueckoo/cmd/importpr.go (hunk 2 above)
Comment (myitcv) [reply cmd/cueckoo/cmd/importpr.go:157]: True, but the noticeable difference here is that the change-id is now visible during the phase of editing the commit message whereas before it was not. Reason being (at least on my setup) is that the change-id is added by a post-edit commit hook. Hence that commit hook now fires as a result of the auto-edit, and the change-id appears in the editor during the final-phase edit.

File: cmd/cueckoo/cmd/importpr.go (hunk 4 above)
Comment (myitcv) [reply cmd/cueckoo/cmd/importpr.go:226]: Thanks for the suggestion. Made this change.

File: cmd/cueckoo/cmd/importpr.go (hunk 2 above)
Comment (myitcv) [reply cmd/cueckoo/cmd/importpr.go:157]: I've fixed the comment to be correct. I mentioned pre-commit, but it's actually post-commit.

**Review by mvdan — APPROVED** [review]

**Comment by mvdan** [comment]
FYI you could also slightly simplify your code by replacing the buffer and calls to Run and String with https://pkg.go.dev/os/exec#Cmd.Output

**Comment by myitcv** [comment]
> FYI you could also slightly simplify your code by replacing the buffer and calls to Run and String with https://pkg.go.dev/os/exec#Cmd.Output

Done. Thanks!

**Pull request merged**